src/VisualStudio/Core/Def/Implementation/Diagnostics/SolutionCrawlerEventHandlerWithDelay.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics
+{
+    internal sealed partial class TaskCenterSolutionAnalysisProgressReporter
+    {
+        /// <summary>
+        /// Helper class to pass on solution crawler events throttled with a certain delay.
+        /// arrive within a certain delay.
+        /// </summary>
+        private class SolutionCrawlerEventHandlerWithDelay
+        {
+            private readonly TimeSpan _publishInterval;
+            private readonly object _lock = new object();
+
+            private ProgressData _progressData;
+            private ProgressData _lastPublishedProgressData;
+
+            /// <summary>
+            /// Task used to invoke events outside the <see cref="_lock"/>
+            /// </summary>
+            private Task _invokeEventTask;
+
+            /// <summary>
+            /// Task used to trigger event invocation on an interval
+            /// defined by <see cref="_publishInterval"/>
+            /// </summary>
+            private Task _intervalTask;
+
+            private bool _isTimerRunning;
```

Comment:
this all seem so complex.  why can't this just be the simple pattern we use elsewhere of:

```c#
Task task;
ProgressData lastProgressData;
object gate = new object();

void SomeEvent(ProgressData data)
{
     lock (gate)
     {
         lastProgressData = data;

         // check if we have an outstanding task that will pick this value up and report it.
         if (task != null)
             return;

         task = Task.Delay(whateverDelay).ContinueWith(ReportProgress);
     }
}

void ReportProgress()
{
  ProgressData data;
  lock (gate)
  {
    data = lastProgressData;
    task = null;
  }    

  UpdateUIWithProgressData(data);
}
```

??

Reply:
benefits of this approach:

1. no need for timers. 
2. code that comes in drops of the latest data for the UI to show.  it then either sees if there's an outstanding task to report that data, or it spawns that task.
3. UI update thread easily grabs the latest data and then marks that any future updates need to trigger the ui after a delay.

It's simple, easy, and works great for both ensuring the UI isn't overloaded (i.e. we throttle with some delay), while also ensuring that any updates to the ui update with latest values we know.

Reply:
this seems workable, not sure why I didn't try it before.  I will try!